### PR TITLE
Regex based sanitizer implementation for logging client and service

### PR DESCRIPTION
**Pull request by amitvc**
The changes in this PR are for [ticket](https://github.com/line/armeria/issues/1810).

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/logging/AbstractSanitizerBase.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.logging;
+import java.util.function.Function;
+
+/**
+ * Base class.
+ */
+public class AbstractSanitizerBase implements Function<Object, Object> {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/AbstractSanitizerBase.java:22]: I guess we don't really need this type. Could you remove this class and just call `toString()` in `RegexBasedSanitizer`?

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Regex based sanitizer.
+ */
+public class RegexBasedSanitizer extends AbstractSanitizerBase implements Function<Object, Object> {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:27]: - `final`
- `Function<Object, String>` ?

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                rawData = m.replaceAll("");
+            }
+        }
+        return rawData;
+    }
+
+    /**
+     * Builder facilitates building regexsanitizer.
+     */
+    public static class RegexBasedSanitizerBuilder {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:54]: If `List<Pattern>` is the only parameter, how about removing this and adding `RegexBasedSanitizer.of(Pattern...)` and `of(Iterable<Pattern>)` ?

File: core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java [hunk 4]
```diff
@@ -95,4 +99,59 @@ void logger() throws Exception {
...
+                             .successfulResponseLogLevel(LogLevel.INFO)
+                             .requestHeadersSanitizer(new RegexBasedSanitizerBuilder()
+                                                              .pattern("trustin")
+                                                              .pattern("com")
+                                                              .build())
+                             .build(delegate);
+
+        // Pre sanitize step
+        Assertions.assertTrue(ctx.logBuilder().toString().contains("trustin"));
+        Assertions.assertTrue(ctx.logBuilder().toString().contains("test.com"));
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java:124]: Sorry, but we use AssertJ for assertions. Could you convert to AssertJ? e.g.

    assertThat(ctx.logBuilder().toString()).contains("trustin");
    ...
    assertThat(ctx.logBuilder().toString()).doesNotContain("trustin");

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java [hunk 5]
```diff
@@ -349,6 +352,64 @@ void sanitizeExceptionIntoString() throws Exception {
...
+                              .requestLogLevel(LogLevel.INFO)
+                              .successfulResponseLogLevel(LogLevel.INFO)
+                              .requestHeadersSanitizer(new RegexBasedSanitizerBuilder()
+                                                               .pattern("trustin")
+                                                               .pattern("com")
+                                                               .build())
+                              .newDecorator().apply(delegate);
+
+        Assertions.assertTrue(ctx.logBuilder().toString().contains("trustin"));
+        Assertions.assertTrue(ctx.logBuilder().toString().contains("test.com"));
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java:380]: Ditto - AssertJ

**Comment by trustin** [comment]
@snaiper80 I think you requested this feature some time ago, right? (My memory might be incorrect, but just in case :sweat_smile:)

**Comment by snaiper80** [comment]
@trustin maybe not. haha. anyway, this pr looks interesting to me. it might be helpful to make a gateway server.

**Comment by trustin** [comment]
Ah, it was @junyoung-ju. :sweat_smile:

File: core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java (hunk 4 above)
Comment (amitvc) [reply core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java]: Done

File: core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java (hunk 5 above)
Comment (amitvc) [reply core/src/test/java/com/linecorp/armeria/server/logging/LoggingServiceTest.java]: Done

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java (hunk 3 above)
Comment (amitvc) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java]: Done

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java (hunk 2 above)
Comment (amitvc) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java]: Made adjustments.

File: core/src/main/java/com/linecorp/armeria/common/logging/AbstractSanitizerBase.java (hunk 1 above)
Comment (amitvc) [reply core/src/main/java/com/linecorp/armeria/common/logging/AbstractSanitizerBase.java]: I removed it.

**Comment by amitvc** [comment]
@trustin Made suggested changes. Ready for review.

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        return rawData;
+    }
+
+    /**
+     * Utility method to create RegexBasedSanitizer.
+     * @param p Pattern.
+     * @return
+     */
+    public static RegexBasedSanitizer of(Pattern...p) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:56]: nit: please check not-null
```suggestion
    public static RegexBasedSanitizer of(Pattern... patterns) {
        requireNonnull("patterns", patterns);   
```

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:56]: Could you add an additaional factory method for `Iterable<? extends Pattern> patterns)`
```
    public static RegexBasedSanitizer of(Iterable<? extends Pattern> patterns);
```

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                rawData = m.replaceAll("");
+            }
+        }
+        return rawData;
+    }
+
+    /**
+     * Utility method to create RegexBasedSanitizer.
+     * @param p Pattern.
+     * @return
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:54]: Could remove two lines.

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Regex based sanitizer.
+ */
+public final class RegexBasedSanitizer implements Function<Object, String> {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:27]: `Function<Object, String>` -> `Function<HttpHeaders, String>`?

File: core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java [hunk 9]
```diff
@@ -95,4 +99,57 @@ void logger() throws Exception {
...
+        final LoggingClient defaultLoggerClient =
+                LoggingClient.builder()
+                             .requestLogLevel(LogLevel.INFO)
+                             .successfulResponseLogLevel(LogLevel.INFO)
+                             .requestContentSanitizer(RegexBasedSanitizer.of(
+                                     Pattern.compile("\\d{3}[-\\.\\s]\\d{3}[-\\.\\s]\\d{4}")))
+                             .build(delegate);
+
+        // Before sanitize content
+        assertThat(ctx.logBuilder().toString().contains("333-490-4499"));
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java:150]:
```suggestion
        assertThat(ctx.logBuilder().toString()).contains("333-490-4499");
```

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java (hunk 6 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:56]: `Pattern` is a final class, so `Iterable<Pattern>` would be enough.

File: core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java [hunk 10]
```diff
@@ -95,4 +99,57 @@ void logger() throws Exception {
...
+        defaultLoggerClient.execute(ctx, req);
+        // Ensure sanitize the request content of the phone number 333-490-4499
+        assertThat(!ctx.logBuilder().toString().contains("333-490-4499"));
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java:153]:
```suggestion
        assertThat(ctx.logBuilder().toString()).doesNotContain("333-490-4499");
```

File: core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java [hunk 11]
```diff
@@ -95,4 +99,57 @@ void logger() throws Exception {
...
+                LoggingClient.builder()
+                             .requestLogLevel(LogLevel.INFO)
+                             .successfulResponseLogLevel(LogLevel.INFO)
+                             .requestHeadersSanitizer(RegexBasedSanitizer.of(
+                                     Pattern.compile("trustin"),
+                                     Pattern.compile("com")))
+                             .build(delegate);
+
+        // Pre sanitize step
+        assertThat(ctx.logBuilder().toString().contains("trustin"));
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/logging/LoggingClientTest.java:122]: `assertThat(ctx.logBuilder().toString()).contains("trustin");`

**Comment by amitvc** [comment]
I will make the suggested changes tonight. Thanks for the review guys.

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java (hunk 8 above)
Comment (amitvc) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:27]: I think we want to be able to use the sanitizers on either httpheader or request content.

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java (hunk 7 above)
Comment (amitvc) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java]: Done

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java (hunk 8 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:30]: Ah, right. Sorry about the comment. 😅

**Comment by amitvc** [comment]
@trustin @ikhoon @minwoox Thank you for the review. I am ready to pick up another issue. Does this PR gets merged when you cut the release?

File: core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    @Override
+    public String apply(@Nullable Object input) {
+        if (input == null) {
+            return null;
+        }
+
+        String rawData = input.toString();
+        for (Pattern pattern : patterns) {
+            final Matcher m = pattern.matcher(rawData);
+            if (m.find()) {
```
Comment (anuraaga) [inline core/src/main/java/com/linecorp/armeria/common/logging/RegexBasedSanitizer.java:70]: `replaceAll` calls `find` internally so I think we can remove this, or it has to run the regex from the beginning again.

**Pull request merged**